feat(storage): add Qdrant support

generall opened the pull request:
Open questions:

- work with offsets
- creation of a new collection on select operations
- Distances
- Mixin inter-dependencies

alaeddine-13 commented on docarray/array/storage/qdrant/getsetdel.py:
```diff
@@ -0,0 +1,118 @@
+                name=self.collection_name,
+                scroll_request=ScrollRequest(
+                    offset=offset,
+                    limit=SCROLL_BATCH_SIZE,
```
can we add this to the config instead ?

alaeddine-13 commented on docarray/array/storage/qdrant/getsetdel.py:
```diff
@@ -0,0 +1,118 @@
+
+from docarray import Document
+from docarray.array.storage.base.getsetdel import BaseGetSetDelMixin
+from docarray.types import ArrayType
```
```suggestion
if TYPE_CHECKING:
    from docarray.types import ArrayType
```

alaeddine-13 commented on docarray/array/storage/qdrant/__init__.py:
```diff
@@ -0,0 +1,42 @@
+    def config(self):
+        return self._config
+
+    def client(self) -> QdrantClient:
```
```suggestion
    @property
    def client(self) -> QdrantClient:
```

alaeddine-13 reviewed (dismissed):
Thanks for your contribution!
I submitted some suggestions.
We also need to add qdrant to `docarray/array/document.py` to enable this syntax:
`da = DocumentArray(storage='qdrant')`
Furthermore, in order to ensure the quality of the Qdrant DocumentArray, tests supported should be activated for Qdrant. We can add a `start_qdrant` fixture in tests/unit/array/conftest.py to start qdrant docker image on module level. Or we can add it to the same docker compose file as weaviate and rename the fixture from start_weaviate to `start_storage`

Find method will be added soon for storage backends that support ANN, it would be great to add this to qdrant as well

alaeddine-13 commented on docarray/array/storage/qdrant/backend.py:
```diff
@@ -0,0 +1,100 @@
+    distance: Distance = field(default=Distance.COSINE)
+    collection_name: Optional[str] = None
+    connection: Optional[Union[str, QdrantClient]] = field(default="localhost:6333")
+    n_dim: Optional[int] = field(default=1)
```
let's raise a `ValueError`  in that case

alaeddine-13 commented on docarray/array/storage/qdrant/getsetdel.py:
```diff
@@ -0,0 +1,118 @@
+
+    def _document_to_qdrant(self, doc: Document, _id: str = None) -> dict:
+        return dict(
+            id=_id or doc.id,
```
Alright, the current behaviour of DocumentArray is to rely on `value.id` as the new ID and `_id` as the existing, old ID.
So basicly, in case both IDs are different, this should change the ID of the Document in the storage backend.

numb3r3 commented on docarray/array/storage/qdrant/getsetdel.py:
```diff
@@ -0,0 +1,122 @@
+
+class GetSetDelMixin(BaseGetSetDelMixin):
+
+    @property
+    def client(self) -> QdrantClient:
+        raise NotImplementedError()
+
+    @property
+    def serialization_config(self) -> dict:
+        raise NotImplementedError()
+
+    @property
+    def n_dim(self) -> int:
+        raise NotImplementedError()
+
+    @property
+    def collection_name(self) -> str:
+        raise NotImplementedError()
+
+    @property
+    def scroll_batch_size(self) -> int:
+        raise NotImplementedError()
```
are these functions are necessary? I found they are implemented in `StorageMixins`

numb3r3 commented on docarray/array/storage/qdrant/seqlike.py:
```diff
@@ -0,0 +1,67 @@
+class SequenceLikeMixin(MutableSequence[Document]):
+
+    @property
+    def client(self) -> QdrantClient:
+        raise NotImplementedError()
+
+    @property
+    def collection_name(self) -> str:
+        raise NotImplementedError()
+
+    @property
+    def config(self):
+        raise NotImplementedError()
+
```
same

generall commented:
Fixed:
- wrong result processing in `FindMixin`
- increase number of open files limit for tests: there are a lot of collections created, should not be required regular work
- Implemented data uploaded with batches - there is 32mb limit for single request, uploading of images should be splitted into multiple batches

JoanFM commented on docarray/array/storage/qdrant/helper.py:
```diff
@@ -22,4 +24,6 @@ def embedding_to_array(
             embedding = np.asarray(embedding).squeeze()
 
+        if np.all(embedding == 0):
+            embedding = embedding + EPSILON
```
Isnt this a little expensive? Also in some cases embeddings are not even numpy arrays and may live in GPU

alaeddine-13 replied:
Qdrant does not accept null vectors because cosine distance can't be computed
Basicly we have 2 options:

- prohibit users from using null vectors as embedding
- implicitly add Epsilon to allow computation
your thoughts ?

numb3r3 replied:
I prefer the second one, as same as we have done in weaviate.

JoanFM commented on docarray/array/storage/qdrant/find.py:
```diff
@@ -0,0 +1,92 @@
+
+
+class FindMixin:
+    @property
```
I think if u declare as abstract method it will not allow u to instantiate the class and will fail faster.

alaeddine-13 replied:
done
